<a href="https://colab.research.google.com/github/leoscarlato/analise_txt_repo/blob/main/atv_scraping.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ATV Scraping - Atividade para Entrega

Nesta atividade, iremos extrair informações de páginas Web. Estes exercícios são para entrega e contam como uma atividade (**ATV**).

## Insper Autograding

Iremos utilizar uma ferramenta para correção automática. Ao responder os exercícios, será possível enviar a solução para um servidor, que dará feedback quase que instantâneo sobre a resposta.

Além de já ter uma boa ideia sobre seu desempenho na atividade, não será necessário fazer a entrega do notebook pelo Blackboard pois o corretor já armazena suas soluções.

Siga os passos deste notebook para realizar a instalação da biblioteca de correção de exercícios nos notebooks da disciplina!

### Instalação

Vamos instalar as bibliotecas:

In [1]:
!python -m pip install -U git+https://github.com/macielcalebe/insperautograding.git

  Cloning https://github.com/macielcalebe/insperautograding.git to /tmp/pip-req-build-rc3jrqt5
  Running command git clone --filter=blob:none --quiet https://github.com/macielcalebe/insperautograding.git /tmp/pip-req-build-rc3jrqt5
  Resolved https://github.com/macielcalebe/insperautograding.git to commit 0d68e50930abe8460efbb43864132ffec30c862b
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 19.2 MB/s eta 0:00:00
  Created wheel for insperautograder: filename=insperautograder-0.2.2-py3-none-any.whl size=4456 sha256=4b053e01a834afa07dce01141f78e7052c5fb05fa0be551a2798bd6e43a10634
  Stored in directory: /tmp/pip-ephem-wheel-cache-0937eebw/wheels/bb/01/cc/7f397587bfedd6c633e2627496c45d5693aa959ee8237ead1f
Successfully built insperautograder


## Importando as bibliotecas

In [2]:
import os
import insperautograder.jupyter as ia

## Configurar ambiente

Execute a célula para configurar nosso ambiente de execução:

In [3]:
os.environ["IAG_OFFERING"] = "25-1"
os.environ["IAG_SUBJECT"] = "analisetxt"
os.environ["IAG_SERVER_URL"] = "https://bigdata.insper-comp.com.br/iag"

## Me diga quem você é

Utilize o **Token** enviado para seu e-mail para que você seja identificado pelo servidor de testes.

Substitua `iagtok_xxx...` pelo token enviado para seu e-mail.

In [4]:
os.environ["IAG_TOKEN"] = "iagtok_6d696baf4b1f66712d83edfa1423148d8b8f50d4d74e8460"

## Conferir atividades e notas

Conferindo quais atividades estão disponíveis!

In [5]:
ia.tasks()

|    | Atividade   | De                  | Até                 |
|---:|:------------|:--------------------|:--------------------|
|  0 | pdf         | 2025-02-13 00:00:00 | 2025-03-20 23:59:59 |
|  1 | scraping    | 2025-03-20 00:00:00 | 2025-04-24 23:59:59 |

Conferindo a nota por exercício na atividade:

In [6]:
ia.grades(task="scraping")

|    | Atividade   | Exercício   |   Peso |   Nota |
|---:|:------------|:------------|-------:|-------:|
|  0 | scraping    | ex01        |      2 |     10 |
|  1 | scraping    | ex02        |      2 |     10 |
|  2 | scraping    | ex03        |      2 |     10 |
|  3 | scraping    | ex04        |      2 |     10 |
|  4 | scraping    | ex05        |      2 |     10 |
|  5 | scraping    | ex06        |      2 |     10 |
|  6 | scraping    | ex07        |      2 |     10 |
|  7 | scraping    | ex08        |      2 |      0 |
|  8 | scraping    | ex09        |      3 |      0 |
|  9 | scraping    | ex10        |      3 |      0 |
| 10 | scraping    | ex11        |      4 |      0 |

Conferindo a média em cada atividade:

In [7]:
ia.grades(by="TASK")

|    | Tarefa   |   Nota |
|---:|:---------|-------:|
|  0 | pdf      |  10    |
|  1 | scraping |   5.38 |

## Exercícios

Vamos praticar os conteúdos das aulas de Scraping!

Responda os exercício e, quando estiver bastante confiante que a resposta está correta, faça o envio para o servidor clicando no botão indicado para isto. Faça correções na sua solução caso o feedback indique erros.

Confira, nas chamadas acima, se a sua nota é atualizada.

Instalações necessárias:

In [8]:
!python -m pip install BeautifulSoup4==4.13.3
!python -m pip install scikit-learn==1.6.1
!python -m pip install selenium==4.29.0
!python -m pip install webdriver_manager==4.0.2

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 57.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 492.9/492.9 kB 32.1 MB/s eta 0:00:00


Vamos importar as bibliotecas:

In [9]:
import requests

### Exercício 1

Crie uma função em Python que receba a URL de uma página (como string). Sua função deve baixar a página com `requests` e **retornar o código HTML completo** da página.

Exemplo de URL:
- https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/aviso.html
- https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem um import como este: `import requests`.
- Considere que a página utiliza como *encoding* o `utf-8`.

In [10]:
# Altere esta função
def baixar_pagina(url):
  response = requests.get(url)
  response.encoding = "utf-8"
  html = response.text
  return html


baixar_pagina("https://www.google.com/url?q=https%3A%2F%2Fatd-insper.s3.us-east-2.amazonaws.com%2Fatv_scraping%2Faviso.html")

'<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"><title>Redirect Notice</title><style>body,div,a{font-family:sans-serif}body{background-color:#fff;margin-top:3px}div{color:#000}a:link{color:#4b11a8;}a:visited{color:#4b11a8;}a:active{color:#ea4335}div.mymGo{border-top:1px solid #dadce0;border-bottom:1px solid #dadce0;background:#f8f9fa;margin-top:1em;width:100%}div.aXgaGb{padding:0.5em 0;margin-left:10px}div.fTk7vd{margin-left:35px;margin-top:35px}</style></head><body><div class="mymGo"><div class="aXgaGb"><font style="font-size:larger"><b>Redirect Notice</b></font></div></div><div class="fTk7vd">&nbsp;The previous page is sending you to <a href="https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/aviso.html">https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/aviso.html</a>.<br><br>&nbsp;If you do not want to visit that page, you can <a href="#" id="tsuid_EZrhZ9yaEqO5wN4P0OS-EA_1">return to the previous page</a>.<script nonce=

Vamos conferir um exemplo de execução:

In [11]:
print(baixar_pagina("https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/aviso.html"))

<!DOCTYPE html>
<html lang="pt-BR">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Aviso - Curso</title>
</head>
<body>
    <h1>Aviso</h1>
    <p><strong>Não haverá aula</strong> durante o período de provas!</p>
</body>
</html>



Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [12]:
ia.sender(answer="baixar_pagina", task="scraping", question="ex01", answer_type="pycode")

interactive(children=(Button(description='Enviar ex01', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 2

Considere, como exemplo, a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html)


Crie uma função em Python que recebe como entrada:
- a **URL** (endereço) de uma página.

A sua função deve:
1) Utilizar a função do exercício anterior para baixar a página. Não precisa enviar a função `baixar_pagina` na resposta, ela já está disponível no servidor!
2) Extrair e retornar, utilizando `BeautifulSoup`, o título da postagem do blog fictício!

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`.

In [13]:
print(baixar_pagina("https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html"))

<!DOCTYPE html>
<html lang="pt">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Post de Blog - Bons Hábitos de Alimentação</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            background-color: #f4f4f4;
            color: #333;
            margin: 0;
            padding: 0;
        }

        header {
            background-color: #4CAF50;
            color: white;
            padding: 10px 0;
            text-align: center;
        }

        article {
            margin: 20px;
            padding: 15px;
            background-color: white;
            border-radius: 5px;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
        }

        h1 {
            font-size: 2rem;
        }

        section.comments {
            margin: 20px;
        }

        section.comments h2 {
            font-size: 1.5rem;
            margin-bottom: 10px;
            color: #4CAF50;
     

In [14]:
import bs4
from bs4 import BeautifulSoup

# Altere esta função
def extrair_titulo_post(url):
  html = baixar_pagina(url)
  soup = BeautifulSoup(html, 'html.parser')
  title = soup.find('h2').text
  return title


Veja um exemplo de execução da função:

In [15]:
url = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html"
extrair_titulo_post(url)

'Como Adotar uma Alimentação Saudável'

Confira se obteve o resultado correto:

```python
"Como Adotar uma Alimentação Saudável"
```

Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [16]:
url = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html"

esperado = "Como Adotar uma Alimentação Saudável"
obtido = extrair_titulo_post(url)

assert  obtido == esperado, f"\nEsperava: {esperado}\n\nObteve: {obtido}"

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [17]:
ia.sender(answer="extrair_titulo_post", task="scraping", question="ex02", answer_type="pycode")

interactive(children=(Button(description='Enviar ex02', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 3

Considere, como exemplo, a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html)


Crie uma função em Python que recebe como entrada:
- O código **HTML** de uma página.

A sua função deve:
- Extrair, utilizando `BeautifulSoup`:
    - O **título** da postagem do blog fictício!
    - O **texto** da postagem.
- Retorne em formato de dicionário. Por exemplo:
    ```python
    {
        "titulo": "Como Adotar uma Alimentação Saudável",
        "texto": "Uma alimentação balanceada é fundamental para manter o corpo saudável e cheio de energia. Neste post, vamos discutir a importância de incluir mais frutas, vegetais e grãos integrais na sua dieta, além de dicas práticas para melhorar sua alimentação."
    }

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`.

In [18]:
# Altere esta função
import bs4
from bs4 import BeautifulSoup

def extrair_json_post(html):
    soup = BeautifulSoup(html, 'html.parser')
    title = soup.find('h2').text
    texto = soup.find('p').text

    final_dict = {
        "titulo" : title,
        "texto": texto
    }

    return final_dict

Veja um exemplo de execução da função (o resultado esperado é `698`)

In [19]:
# Vamos conseguir um HTML de exemplo
url = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html"
html = baixar_pagina(url)

# Agora chamamos a função que você implementou
extrair_json_post(html)

{'titulo': 'Como Adotar uma Alimentação Saudável',
 'texto': 'Uma alimentação balanceada é fundamental para manter o corpo saudável e cheio de energia. Neste post, vamos discutir a importância de incluir mais frutas, vegetais e grãos integrais na sua dieta, além de dicas práticas para melhorar sua alimentação.'}

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [20]:
ia.sender(answer="extrair_json_post", task="scraping", question="ex03", answer_type="pycode")

interactive(children=(Button(description='Enviar ex03', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 4

De forma semelhante ao exercício anterior, considere a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/istoedinheiro/ouro.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/istoedinheiro/ouro.html)

Crie uma função em Python que recebe como entrada:
- O código **HTML** de uma página da **IstoÉDinheiro**.

A sua função deve:
- Extrair, utilizando `BeautifulSoup`:
    - O **título** da notícia!
    - O **primeiro parágrafo** da notícia.
    - A **data** de publicação da notícia.
- Retorne em formato de dicionário. Por exemplo:
    ```python
    {
        "titulo": "...",
        "primeiro_paragrafo": "..."
        "data": "..."
    }

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`.

In [21]:
# Altere esta função
import bs4
from bs4 import BeautifulSoup

def extrair_json_isto_e(html):
    soup = BeautifulSoup(html, 'html.parser')
    artigo = soup.find('article')
    title = soup.find('h1').text
    primeiro_paragrafo = artigo.find_all('p')[2].text
    data = artigo.find_all('p')[1].text

    final_dict = {
        "titulo":title,
        "primeiro_paragrafo":primeiro_paragrafo,
        "data":data
    }

    return final_dict

Veja um exemplo de execução da função (o resultado esperado é `698`)

In [22]:
# Vamos conseguir um HTML de exemplo
url_isto_e = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/istoedinheiro/ouro.html"
html_isto_e = baixar_pagina(url_isto_e)

# Agora chamamos a função que você implementou
extrair_json_isto_e(html_isto_e)

{'titulo': 'Ouro atinge novo recorde histórico impulsionado por demanda por ativos de refúgio',
 'primeiro_paragrafo': 'O\n ouro teve uma variação limitada nesta quarta-feira, 19, ainda que tenha\n atingido um novo recorde histórico impulsionado pela demanda por ativos\n de refúgio, tensões geopolíticas e compras persistentes do metal por \nbancos centrais. Com o Federal Reserve (Fed, o banco central \nnorte-americano) prestes a anunciar sua decisão sobre política \nmonetária, o mercado acompanha de perto os próximos desdobramentos.',
 'data': '19/03/2025 - 15:07'}

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [23]:
ia.sender(answer="extrair_json_isto_e", task="scraping", question="ex04", answer_type="pycode")

interactive(children=(Button(description='Enviar ex04', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 5

Considere, como exemplo, a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_ciclismo.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_ciclismo.html)


Crie uma função em Python que recebe como entrada:
- O código **HTML** de uma página.

A sua função deve:
- Contar e retornar, como inteiro, a quantidade de comentários na página.

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`.

In [24]:
# Altere esta função
import bs4
from bs4 import BeautifulSoup

def contar_comentarios(html):
  soup = BeautifulSoup(html, 'html.parser')
  comments = soup.find('section', {'class':'comments'})
  all_comments = len(comments.find_all('div', {'class':'comment'}))
  return all_comments

Veja um exemplo de execução:

In [25]:
# Vamos conseguir um HTML de exemplo
url = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html"
html = baixar_pagina(url)

# Agora chamamos a função que você implementou
qt_comentarios = contar_comentarios(html)

print(f"Foram encontrados {qt_comentarios} comentários")

Foram encontrados 4 comentários


Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [26]:
assert contar_comentarios(html) == 4

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [27]:
ia.sender(answer="contar_comentarios", task="scraping", question="ex05", answer_type="pycode")

interactive(children=(Button(description='Enviar ex05', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 6

Considere, como exemplo, a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_ciclismo.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_ciclismo.html)


Crie uma função em Python que recebe como entrada:
- O código **HTML** de uma página.

A sua função deve:
- Retornar uma **lista** com o nome de todos os usuários que realizaram comentários!

Responda alterando o código da função abaixo.

Obs:

- Retorne os usuários na ordem em que comentaram. O nome do usuário será repetido caso tenha efetuado mais de um comentário (e tudo bem).
- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`.

In [28]:
# Altere esta função
import bs4
from bs4 import BeautifulSoup

def get_usuarios(html):
  soup = BeautifulSoup(html, 'html.parser')
  comments = soup.find_all('div', {'class':'comment'})
  all_names = []
  for c in comments:
    all_names.append(c.find('strong').text.replace(":", ""))

  return all_names

Veja um exemplo da execução da função:

In [29]:
# Vamos conseguir um HTML de exemplo
url = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html"
html = baixar_pagina(url)

# Agora chamamos a função que você implementou
usuarios = get_usuarios(html)

print(f"Usuários: {usuarios}")

Usuários: ['João', 'Ana', 'Carlos', 'Mariana']


É esperado que ela retorne:


```python
['João', 'Ana', 'Carlos', 'Mariana']
```


Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [30]:
ia.sender(answer="get_usuarios", task="scraping", question="ex06", answer_type="pycode")

interactive(children=(Button(description='Enviar ex06', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 7

Considere, como exemplo, a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_ciclismo.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_ciclismo.html)


Crie uma função em Python que recebe como entrada:
- O código **HTML** de uma página.

A sua função deve:
- Retornar um **DataFrame** pandas contendo os **usuários** e seus **comentários**!

Responda alterando o código da função abaixo.

Obs:

- As colunas devem ser `usuario` e `comentario`, sem acentuação.
- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`. Ainda, já tem `import pandas as pd`.

In [31]:
# Altere esta função
import bs4
from bs4 import BeautifulSoup
import pandas as pd

def get_df_comentarios(html):
  soup = BeautifulSoup(html, 'html.parser')
  comments = soup.find_all('div', {'class':'comment'})
  all_names = []
  all_comments = []
  for c in comments:
    all_names.append(c.find('strong').text.replace(":", ""))
    all_comments.append(c.find('p').text.split(":")[1].strip())

  df = pd.DataFrame()
  df['usuario'] = all_names
  df['comentario'] = all_comments
  return df

Veja um exemplo da execução da função:

In [32]:
# Vamos conseguir um HTML de exemplo
url = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/comentarios_alimentacao.html"
html = baixar_pagina(url)

# Agora chamamos a função que você implementou
df_comentarios = get_df_comentarios(html)

df_comentarios

usuario                                         comentario
0     João  Ótimas dicas! Já estou tentando incluir mais v...
1      Ana  Esse post me motivou a começar a fazer smoothi...
2   Carlos  A alimentação saudável realmente faz toda a di...
3  Mariana  Dicas excelentes! Vou começar a substituir os ...

É esperado que ela retorne:

|    | usuario   | comentario                                                                                                             |
|---:|:----------|:-----------------------------------------------------------------------------------------------------------------------|
|  0 | João      | Ótimas dicas! Já estou tentando incluir mais vegetais nas minhas refeições diárias.                                    |
|  1 | Ana       | Esse post me motivou a começar a fazer smoothies com frutas! Muito obrigada!                                           |
|  2 | Carlos    | A alimentação saudável realmente faz toda a diferença no dia a dia. Vou começar a planejar melhor as minhas refeições. |
|  3 | Mariana   | Dicas excelentes! Vou começar a substituir os alimentos processados por opções mais naturais.                          |

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [33]:
ia.sender(answer="get_df_comentarios", task="scraping", question="ex07", answer_type="pycode")

interactive(children=(Button(description='Enviar ex07', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 8

Considere, como exemplo, a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/produtos_v1.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/produtos_v1.html)


Crie uma função em Python que recebe como entrada:
- O código **HTML** de uma página.

A sua função deve:
- Contar e retornar, como inteiro, a quantidade de **produtos** na página.

Responda alterando o código da função abaixo.

Obs:

- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`.

In [42]:
# Altere esta função
import bs4
from bs4 import BeautifulSoup

def contar_produtos(html):
  soup = BeautifulSoup(html, 'html.parser')
  products = soup.find_all('div', {'class':'product'})
  return len(products)

Veja um exemplo de execução:

In [43]:
# Vamos conseguir um HTML de exemplo
url_prod = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/produtos_v1.html"
html_prod = baixar_pagina(url_prod)

# Agora chamamos a função que você implementou
qt_produtos = contar_produtos(html_prod)

print(f"Foram encontrados {qt_produtos} produtos")

Foram encontrados 3 produtos


Abaixo, seguem alguns testes que rodam localmente e verificam sua resposta.

In [44]:
assert contar_produtos(html_prod) == 3

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [45]:
ia.sender(answer="contar_produtos", task="scraping", question="ex08", answer_type="pycode")

interactive(children=(Button(description='Enviar ex08', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 9

Considere, como exemplo, a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/produtos_v1.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/produtos_v1.html)


Crie uma função em Python que recebe como entrada:
- O código **HTML** de uma página.

A sua função deve:
- Retornar um **DataFrame** pandas contendo os **produtos**, suas **descrições** e **preços**.

Responda alterando o código da função abaixo.

Obs:

- As colunas devem ser `produto`, `descricao` e `preco`, sem acentuação.
- O preço deve ser limpo e transformado para `float`.
- O nome e argumentos da função não devem ser alterados.
- Considere que o código de correção já tem o import da `bs4` e `BeautifulSoup`. Ainda, já tem `import pandas as pd`.

In [80]:
# Altere esta função
import bs4
from bs4 import BeautifulSoup

def get_df_produtos(html):
  soup = BeautifulSoup(html, 'html.parser')
  products = soup.find_all('div', {'class':'product'})
  all_names = []
  all_desc = []
  all_prices = []
  for p in products:
    all_names.append(p.find('h2').text)
    all_desc.append(p.find('p').text)
    all_prices.append(float(p.find('p', {'class':'price'}).text.replace("R$", "").replace(",", ".").strip()))

  df = pd.DataFrame()
  df['produto'] = all_names
  df['descricao'] = all_desc
  df['preco'] = all_prices

  return df


Veja um exemplo da execução da função:

In [81]:
# Vamos conseguir um HTML de exemplo
url_prod = "https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/produtos_v1.html"
html_prod = baixar_pagina(url_prod)

# Agora chamamos a função que você implementou
df_prod = get_df_produtos(html_prod)

df_prod

produto                                          descricao  preco
0  Produto 1  Descrição do Produto 1. Este produto é ótimo p...   99.9
1  Produto 2  Descrição do Produto 2. Ideal para quem busca ...   79.5
2  Produto 3  Descrição do Produto 3. Um produto inovador qu...  149.0

É esperado que ela retorne:

|    | produto   | descricao                                                                                              |   preco |
|---:|:----------|:-------------------------------------------------------------------------------------------------------|--------:|
|  0 | Produto 1 | Descrição do Produto 1. Este produto é ótimo para quem procura qualidade e durabilidade.               |    99.9 |
|  1 | Produto 2 | Descrição do Produto 2. Ideal para quem busca praticidade no dia a dia, com excelente custo-benefício. |    79.5 |
|  2 | Produto 3 | Descrição do Produto 3. Um produto inovador que vai facilitar a sua vida, com características únicas.  |   149.0 |

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [82]:
ia.sender(answer="get_df_produtos", task="scraping", question="ex09", answer_type="pycode")

interactive(children=(Button(description='Enviar ex09', style=ButtonStyle()), Output()), _dom_classes=('widget…

### Exercício 10

Considere a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/dicas.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/dicas.html)

Acesse em seu navegador e aparte **F5** em seu teclado (ou o botão de atualizar no navegador). Você irá perceber que esta página possui **conteúdo dinâmico**, exibindo uma dica diferente a cada atualização.

Assim, iremos utilizar **Selenium** para extrair dicas desta página.

Crie uma função python que recebe um **driver** já instanciado. Sua função deve:

- Utilizar o driver para acessar a página.
- Extrair a dica exibida.

Vamos criar o driver, mas antes, instalaremos com:

In [ ]:
!python -m pip install selenium
!python -m pip install webdriver_manager

Se estiver no Google Colab, descomente e também rode:

In [ ]:
# !sudo apt update
# !wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
# !sudo dpkg -i google-chrome-stable_current_amd64.deb
# !sudo apt-get -f install
# !google-chrome-stable --version

Vamos importar as bibliotecas:

In [ ]:
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By

E finalmente criar o driver! Execute a linha abaixo. Você irá perceber que será criada uma janela do navegador Chrome. Deixe esta janela aberta.

Obs: No Colab, não há como visualizar o navegador utilizado pelo Selenium!

In [ ]:
driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()))
driver.set_page_load_timeout(60)

# Se estiver no Colab, comente o código acima e utilize como alternativa
# chrome_options = webdriver.ChromeOptions()
# chrome_options.add_argument('--headless') # ensure GUI is off
# chrome_options.add_argument('--no-sandbox')
# chrome_options.add_argument('--disable-dev-shm-usage')
# driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=chrome_options)
# driver.set_page_load_timeout(60)

In [ ]:
# Altere esta função
def obter_dica(driver):
    return ""

Veja um exemplo da execução da função:

In [ ]:
print(obter_dica(driver))

print(obter_dica(driver))

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="obter_dica", task="scraping", question="ex10", answer_type="pycode")

### Exercício 11

Considere a página disponível em [https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/dicas_v2.html](https://atd-insper.s3.us-east-2.amazonaws.com/atv_scraping/dicas_v2.html)

Acesse em seu navegador e clique em **Nova Dica**. Você irá perceber que esta página possui **conteúdo dinâmico**, exibindo uma dica diferente a cada clique.

Assim, iremos utilizar **Selenium** para extrair dicas desta página.

Crie uma função python que recebe um **driver** já instanciado. Sua função deve:

- Utilizar o driver para acessar a página.
- Identificar e clicar no botão.
- Extrair a dica exibida.

In [ ]:
# Altere esta função
def obter_dica_com_clique(driver):
    return ""

Veja um exemplo da execução da função:

In [ ]:
print(obter_dica_com_clique(driver))

print(obter_dica_com_clique(driver))

Quando estiver confiante que sua resposta está correta, clique no botão abaixo para fazer o envio.

In [ ]:
ia.sender(answer="obter_dica_com_clique", task="scraping", question="ex11", answer_type="pycode")

## Conferir notas

Vamos verificar se as notas foram gravadas nos exercícios!

In [ ]:
ia.grades(task="scraping")

Vamos verificar se a nota da atividade está ok:

In [ ]:
ia.grades(by="TASK")

## Finalizando o trabalho

Por hoje é só!